### PR TITLE
add-verify-genomic-intervals-to-core

### PR DESCRIPTION
**Pull request by gfudenberg**
relates to https://github.com/open2c/bioframe/issues/48,  "create a new module that defines standards on bioframes, verifies existing dataframes and converts different inputs into bioframes: core.py? bioframe.py? standards.py? utils.py?"

rough draft, but hopefully easier to discuss here than in a gist. also make_region_df() still needs an option to input a pd.Series, or a df with just  ['region','position'], a la current chromsizes.

### CODE REVIEW

**Comment by gfudenberg** [comment]
in addition to new functions in core, this PR also includes:
- big updates to `complement()` & `trim()` (and corresponding changes to tests)
- moving some of the verifiers from ops to core

**Comment by gfudenberg** [comment]
check out discussion points (marked by TODO) in:
- assign_view
    - discuss if we should add a drop_unassigned=True option.
- sanitize_bioframe:  
    - discuss dropping of intervals based on view_df & df_view_col
    - whether to include return_sorted, and if so, how.

File: bioframe/core.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+    if not _verify_columns(
+        region_df, [ck1, sk1, ek1, view_name_col], return_as_bool=True
+    ):
+        if raise_errors:
+            raise TypeError("Invalid view: column names cannot be verified")
+        return False
+
+    if not is_bedframe(region_df, cols=cols):
+        if raise_errors:
+            raise ValueError("Invalid view: bedframe properties cannot be verified")
```
Comment (nvictus) [inline bioframe/core.py:221]: Could chain exceptions here

File: bioframe/core.py [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+
+    if is_overlapping(region_df, cols=cols):
+        if raise_errors:
+            raise ValueError("Invalid view: entries must be non-overlapping")
+        return False
+
+    return True
+
+
+def make_viewframe(
```
Comment (sergpolly) [inline bioframe/core.py:244]: this duplicates functionality of the [`bioframe.parse_regions`](https://github.com/open2c/bioframe/blob/0508270bac18ef15bdea9535d9126c958d6c9952/bioframe/region.py#L160) by @mimakaev  - we should merge these 2 functions, by bringing some of the functionality from `parse_regions` to `make_viewframe`:
 - force UCSC renaming ...
 - fill_missing ?
 - check_bounds ?

Also from the discussion - drop support for `{str:(int,int)}` format and `p/q` stripping ...

File: bioframe/core.py [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+    Returns
+    -------
+    is_tiling:bool
+
+    """
+
+    view_df = make_viewframe(view_df, view_name_col=view_name_col, cols=cols)
+
+    if is_overlapping(df):
+        print("overlaps")
```
Comment (Phlya) [inline bioframe/core.py:458]: Change to (optional) raising

File: bioframe/core.py (hunk 2 above)
Comment (sergpolly) [reply bioframe/core.py:244]: to specify - functionality from `parse_regions` that we actually use in cooltools:
 - make the function check bounds of regions in a view by providing a dict-like for chromsizes
 - add `name` to regions even if they do not have one, by default force UCSC formatting, or allow a formatting function e.g. `lambda chrom, start, end: f"{chrom}:{start}-{end}"`

Not sure about `fill_missing` - never used it

Comment (sergpolly) [reply bioframe/core.py:244]: add legacy format for `view`(`regions`) - `[(chom, start, end)]` - and/or  - `[(chom, start, end, name)]` - potentially copy/paste from `bioframe.parse_regions`

File: bioframe/core.py (hunk 1 above)
Comment (gfudenberg) [reply bioframe/core.py:221]: do you have an example where this is done?

File: bioframe/core.py (hunk 2 above)
Comment (gfudenberg) [reply bioframe/core.py:244]: actually, seems like old check_bounds only check starts < ends... added an option to pass a dictlike chromsizes anyways

**Comment by gfudenberg** [comment]
bioframe.io.formats → bioframe.io.fileops
assign_view, sort_bedframe → ops
Make folder: .core.specs, .core.construction, core.stringops, core.checks
-- column names & dtypes → .core.specs
-- bioframe.region.parse* → .core.stringops
-- bioframe.region.from* → .core.construction
-- bioframe.core → .core.checks (is_*), .core.construction (sanitize, make_viewframe)
-- from_ → from_any
-- add underscore to parse_humanized, deleted atoi

**Pull request merged**